azure/services/networkinterfaces/spec.go

```diff
@@ -80,13 +81,26 @@ func (s *NICSpec) OwnerResourceName() string {
 }
 
 // Parameters returns the parameters for the network interface.
-func (s *NICSpec) Parameters(_ context.Context, existing interface{}) (parameters interface{}, err error) {
+func (s *NICSpec) Parameters(ctx context.Context, existing interface{}) (parameters interface{}, err error) {
+	_, log, done := tele.StartSpanWithLogger(ctx, "networkinterfaces.NICSpec.Parameters")
+	defer done()
+
 	if existing != nil {
-		if _, ok := existing.(armnetwork.Interface); !ok {
+		existingNIC, ok := existing.(armnetwork.Interface)
+		if !ok {
 			return nil, errors.Errorf("%T is not an armnetwork.Interface", existing)
 		}
-		// network interface already exists
-		return nil, nil
+
+		// If the NIC already exists, return a nil parameters and nil errors to skip the create or update of the already existing NIC.
+		// If the NIC is in ProvisioningFailed state, we will try to build the parameters of the existing NIC as the provisioning failed does not necessarily mean the NIC is in a bad state.
+		// Reference: https://learn.microsoft.com/en-us/azure/networking/troubleshoot-failed-state#provisioning-states and
+		// https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/5515
+		if existingNIC.Properties != nil && *existingNIC.Properties.ProvisioningState != armnetwork.ProvisioningStateFailed {
```

Comment:
We need to check for `ProvisioningState *ProvisioningState` nil before we dereference it